### DISCUSSION TITLE
Folder duplication during first run with linux client

### DISCUSSION THREAD

**Discussion by vroby67** [abraunegg/onedrive#2327] [Q&A] [answered]
Hi there,

I'm working on getting a SMB network share on a linux machine synced with onedrive.
I probably made the mistake of working in a single session on:
 - migrating data from ms server 2019 to linux folder, mounting the cifs on linux and running rsync towards the local folder
 - setting up onedrive client on linux system
 - configuring samba in a way to preserve the owner of a file, so giving 664 (or 774) on all the shared files
 - migrating onedrive to docker.
Particularly during the samba configuration I made a lot of tests verifying the attributes propagation from windows client.

Whatever the cause, now I have either locally either on onedrive online some folders duplicated, usually 3 copies with the following schema:
```
drwx------   5 code   code 4.0K Mar  6 18:09 '0516 - Plastlab - Generatore pulsazione idrostatico'
drwxrwxr-x   5 code   code 4.0K Dec 14 11:58 '0516 - PLASTLAB - Generatore pulsazione idrostatico'
drwx------   5 code   code 4.0K Mar  6 18:28 '0516 - PLASTLAB - Generatore pulsazione idrostatico 1'
```
while online I can find
```
0516 - PLASTLAB - Generatore pulsazione idrostatico 1
0516 - Plastlab - Generatore pulsazione idrostatico
```
both marked as new element (at least half an hour after local deletion).

Now I'm wondering on how to fix the situation, hopefully without flashing out onedrive online and restarting from scratch the sync from my last backup.

I tried to be concise in summarizing the situation without getting lost in information that may be unnecessary.
If any willing and interested person would like to delve deeper into the matter, I will be happy to provide all the necessary information deemed necessary.

Thank you all in advance

**Comment by abraunegg** [comment, marked as answer]
@vroby67 
The issue you describe has nothing to do with this client.

If a 'folder' already exists, it will *never* attempt to create it. It checks both the local filesystem and remote before any folder creation is performed.

Additionally, the client adheres to the actual names of the folders including any specific capitalisation. It *does not change* this *ever*. Windows and possibly OneDrive online (and maybe Samba as well) will add a '1' to folder names if the folder already exists.

The situation you are describing sounds like a Windows 2019 or Samba issue, but has 100% nothing to do with this client.

I suggest that you:
* Raise your issue with Microsoft
* Raise your issue with the Samba team
* Enable verbose logging on the 'onedrive' client and have this logging to a logfile so that you can validate what the client is pulling down from OneDrive via the API and what it is sending back.

Sorry to not be much help otherwise - but this is client is 100% not the cause.

**Reply by vroby67** [reply]
Thanks for the reply.
You will agree how puzzling the behavior is. I agree with what you say; the doubts about your client arise from some behaviors that I can't understand: for example, the docker starts only with `--resync` in the configuration file, otherwise it doesn't start (or rather, it keeps restarting) without me seeing messages on the log or in the integrated console. With resync on, it seems to work fine. 

I've read that you repeatedly recommend not turning on resyncing unless explicitly requested by the client, but right now I can't do otherwise. Most importantly, could you please tell me how to diagnose this operating condition?

I know I'm affected by a form of Gundam syndrome (I can't say how serious) but I'm working hard to train myself in the skills to possess the subject and your help would be very important.

I haven't even been able to interpret messages like:
```
2023-Mar-07 07:45:10.2167698    Skipping uploading this new file: /onedrive/data/VBAnet/vbaServer/A - Commesse/0562e - Fortek - Carrello miscelatore miscelatore/Swk/170254_
```
...why was the upload skipped?

or also

```
2023-Mar-07 06:57:51.5126497    Skipping uploading this new file as parent path is not in the database: ./VBAnet/vbaServer/A - Commesse/0219 - FORTEK - Sostituzione motori Pieffeci/Tecnico/Progetto/mR-42-01 - VBA -  block notes.doc
2023-Mar-07 06:57:52.7314917    Skipping uploading this new file as parent path is not in the database: ./VBAnet/vbaServer/A - Commesse/0219 - FORTEK - Sostituzione motori Pieffeci/Tecnico/Foto/Ozzio/IMG_20210518_113246.jpg
```

Is it perhaps a timeline issue causing the file to exist while the folder's existence is not yet registered? If so, the upload could take place as soon as the existence of the folder takes shape in the database...

Thanks anyway, even if you don't think it's appropriate to help me. I understand that this may be too specific an issue to be of interest to the community.
Cheers

**Comment by abraunegg** [comment]
@vroby67 
> the doubts about your client arise from some behaviors that I can't understand: for example, the docker starts only with --resync in the configuration file

Well that is 100% not right - and points me to potentially that you have not installed Docker and/or the client under Docker correctly. The *default* configuration for the client is that it runs without needing resync ... so how you have set things up is most likely a major contributing factor.
 
Stepping through your steps:

> migrating data from ms server 2019 to linux folder, mounting the cifs on linux and running rsync towards the local folder

OK - there should not be a problem here - so long as you are using the right command, and not trying to mount that same Windows Share somewhere else. This should be a a one-way copy. If you have the Windows version of OneDrive running on your server, it would be advisable to disable it.

> setting up onedrive client on linux system

It is unlcear here if you are running the client without Docker or with Docker ... my suggestion here is that because you are having issues with running the client with Docker, strip away complexity and run the client *without* Docker. This way you can control the client easier, configure the client easier, increase the logging verbosity and/or use `--dry-run` to test and validate what you want to do.

> configuring samba in a way to preserve the owner of a file, so giving 664 (or 774) on all the shared files

It is not clear here as to why you are doing this - presumably to then access all your files again rather than from the Windows Server. Doing this has zero impact on how this OneDrive client operates. The caveat here is that the user/group the 'onedrive' client is running as - is part of the 'group' that has read/write access to the files.

> I haven't even been able to interpret messages like: .... as to why items are skipped

See above as to my point about Docker. Something is *not right* with how you have implemented the client.

If the client cannot look up data that should be there - that points to the client not being able to write data into the database, which leads to why you say you need to keep starting the client with resync - because that attempts to clean up the database so you start from scratch.

The next best course of action for you is to remove using Docker. Run the client normally without Docker on your platform. Configure it correctly, test it's execution with `--dry-run` before manually testing its operation. Once you are happy with it, enable to systemd service so that it is running in the background.

Once you are then happy with how that is working, go back to looking at your Docker installation and configuration as you most likely did not follow the instructions correctly there.
